pwiz_tools/Skyline/TestData/ConsoleAddAnnotationsTest.cs

```diff
@@ -0,0 +1,256 @@
+                "--annotation-type=" + "value_list", // Specify the type
+                "--save"
+            );
+            CommandLineTest.CheckRunCommandOutputContains(
+                string.Format(
+                    Resources.CommandLine_AddAnnotationsFromArguments_Error__Cannot_add_a__0__type_annotation_without_providing_a_list_values_of_through__1__,
+                    AnnotationDef.AnnotationType.value_list.ToString(), CommandArgs.ARG_ADD_ANNOTATIONS_VALUES.ArgumentText), output);
+        }
+
+        [TestMethod]
+        public void ConsoleInvalidAnnotationTypeTest()
+        {
+            TestFilesDir = new TestFilesDir(TestContext, @"TestData\ConsoleAddAnnotationsTest.zip");
+            var newDocumentPath = TestFilesDir.GetTestPath("out.sky");
+            // Test error (invalid annotation-type value)
+            var output = RunCommand("--new=" + newDocumentPath, // Create a new document
+                "--overwrite", // Overwrite, as the file may already exist in the bin
+                "--annotation-name=" + ANNOTATION_NAME, // Name the annotation
+                "--annotation-targets=" + ANNOTATION_TARGETS, // Specify the target
+                "--annotation-type=" + INVALID_VALUE // Specify an invalid type value
+            );
+            CommandLineTest.CheckRunCommandOutputContains(string.Format(
+                Resources.ValueInvalidException_ValueInvalidException_The_value___0___is_not_valid_for_the_argument__1___Use_one_of__2_,
+                INVALID_VALUE,
+                "--annotation-type",
+                string.Join(@", ", ListPropertyType.ListPropertyTypes().Select(c =>
+                    c.AnnotationType.ToString()).ToArray())), output);
+        }
+
+        [TestMethod]
+        public void ConsoleAddXmlDefinedAnnotationTest()
+        {
+            var annotationValuesArray = ANNOTATION_VALUES.Split(',');
+            TestFilesDir = new TestFilesDir(TestContext, @"TestData\ConsoleAddAnnotationsTest.zip");
+            var annotationsXml = TestFilesDir.GetTestPath("Annotations.xml");
+            var annotationsXmlBadFormatting = TestFilesDir.GetTestPath("AnnotationsIncorrectFormatting.xml");
+            var newDocumentPath = TestFilesDir.GetTestPath("out.sky");
+            // Test define (from .xml file)
+            var output = RunCommand("--new=" + newDocumentPath, // Create a document (without annotations)
+                "--overwrite", // Overwrite, as the file may already exist in the bin
+                "--annotation-file=" + annotationsXml, // Specify file path
+                "save"
+            );
+            CommandLineTest.CheckRunCommandOutputContains(string.Format(
+                    Resources.CommandLine_AddAnnotations_Annotations_successfully_defined_from_file__0__, annotationsXml)
+                , output);
+            // Assert that the document has the correct number of annotations
+            AssertDocumentAnnotationCount(newDocumentPath, 2);
```

Comment:
I would recommend that you get rid of the methods "AssertDocumentAnnotationCount" and "AssertAnnotationInDocumentAndSettings" and replace them with the following:
```
            var doc = ResultsUtil.DeserializeDocument(newDocumentPath);
            Assert.AreEqual(2, doc.Settings.DataSettings.AnnotationDefs.Count);
            var peptideQualityAnnotation = new AnnotationDef("Peptide Quality",
                AnnotationDef.AnnotationTargetSet.Singleton(AnnotationDef.AnnotationTarget.peptide),
                AnnotationDef.AnnotationType.value_list, annotationValuesArray);
            CollectionAssert.Contains(doc.Settings.DataSettings.AnnotationDefs.ToList(), peptideQualityAnnotation);
            CollectionAssert.Contains(Settings.Default.AnnotationDefList, peptideQualityAnnotation);

            var bioReplicateAnnotation = new AnnotationDef("BioReplicate",
                AnnotationDef.AnnotationTargetSet.Singleton(AnnotationDef.AnnotationTarget.replicate),
                AnnotationDef.AnnotationType.number, null);
            CollectionAssert.Contains(doc.Settings.DataSettings.AnnotationDefs.ToList(), bioReplicateAnnotation);
            CollectionAssert.Contains(Settings.Default.AnnotationDefList, bioReplicateAnnotation);
```
The easiest way to check whether an annotation is in a list is to create an AnnotationDef with the values you are looking for, and then use CollectionAssert.Contains.

When you are able to do that, "AssertAnnotationInDocumentAndSettings" ends up being a simple enough method that it's probably not worth having it as a method. It's simpler to just deserialize the document once, and do the asserts that you want on it.

Also, once you have the document in a local variable, you might think of other things you might want to assert about it.

Reply:
Thanks, that is simpler